core/src/main/scala/org/apache/spark/resource/ResourceProfileManager.scala

```diff
@@ -59,35 +59,64 @@ private[spark] class ResourceProfileManager(sparkConf: SparkConf,
+      }
+    } else {
+      val isNotDefaultProfile = rp.id != ResourceProfile.DEFAULT_RESOURCE_PROFILE_ID
+      val notYarnOrK8sOrStandaloneAndNotDefaultProfile =
+        isNotDefaultProfile && !(isYarn || isK8s || isStandalone)
+      val YarnOrK8sOrStandaloneNotDynAllocAndNotDefaultProfile =
+        isNotDefaultProfile && (isYarn || isK8s || isStandalone) && !dynamicEnabled
+
+      // We want the exception to be thrown only when we are specifically testing for the
+      // exception or in a real application. Otherwise in all other testing scenarios we want
+      // to skip throwing the exception so that we can test in other modes to make testing easier.
+      if ((notRunningUnitTests || testExceptionThrown) &&
         (notYarnOrK8sOrStandaloneAndNotDefaultProfile ||
           YarnOrK8sOrStandaloneNotDynAllocAndNotDefaultProfile)) {
-      throw new SparkException("ResourceProfiles are only supported on YARN and Kubernetes " +
-        "and Standalone with dynamic allocation enabled.")
-    }
+        throw new SparkException("ResourceProfiles are only supported on YARN and Kubernetes " +
+          "and Standalone with dynamic allocation enabled.")
+      }
 
-    if (isStandalone && rp.getExecutorCores.isEmpty &&
-      sparkConf.getOption(config.EXECUTOR_CORES.key).isEmpty) {
-      logWarning("Neither executor cores is set for resource profile, nor spark.executor.cores " +
-        "is explicitly set, you may get more executors allocated than expected. It's recommended " +
-        "to set executor cores explicitly. Please check SPARK-30299 for more details.")
+      if (isStandalone && dynamicEnabled && rp.getExecutorCores.isEmpty &&
+        sparkConf.getOption(config.EXECUTOR_CORES.key).isEmpty) {
+        logWarning("Neither executor cores is set for resource profile, nor spark.executor.cores " +
+          "is explicitly set, you may get more executors allocated than expected. " +
+          "It's recommended to set executor cores explicitly. " +
+          "Please check SPARK-30299 for more details.")
+      }
     }
 
     true
   }
 
+  /**
+   * Check whether a task with specific taskRpId can be scheduled to executors
+   * with executorRpId.
+   *
+   * Here are the rules:
+   * 1. Tasks with [[TaskResourceProfile]] can be scheduled to executors with
+   *    default resource profile when dynamic allocation is disabled.
+   * 2. Other tasks can be scheduled to executors where resource profile exactly matches.
+   */
+  private[spark] def canBeScheduled(taskRpId: Int, executorRpId: Int): Boolean = {
+    assert(resourceProfileIdToResourceProfile.contains(taskRpId) &&
+      resourceProfileIdToResourceProfile.contains(executorRpId),
+      "Tasks and executors must have valid resource profile id")
+    val taskRp = resourceProfileFromId(taskRpId)
+
+    taskRpId == executorRpId || (!dynamicEnabled && taskRp.isInstanceOf[TaskResourceProfile] &&
+      executorRpId == ResourceProfile.DEFAULT_RESOURCE_PROFILE_ID)
```

Comment:
It seems like when creating `TaskResourceProfile` (either by `ResourceProfileBuilder.build()` or manually creating one), `TaskResourceProfile` always has a unique rpId. However, I thought you'd expect to have the constant rpId (i.e., ResourceProfile.DEFAULT_RESOURCE_PROFILE_ID) for `TaskResourceProfile` in the case of dynamic disabled.

Reply:
I would not expect it to have the same id as the default resource profile id because its not he same profile, it may have the same executor resource requirements but the task ones are different.

Also have you tested the UI, how does this show up in the environment page?

Reply:
> I would not expect it to have the same id as the default resource profile id because its not he same profile, 

Agree. So the condition should be updated. It should not compare the rp id but the executor resource requirement.


@ivoson reminder for testing on UI.

Reply:
In the UI page, it shows resource profile id and task resources. Here is an example:
![image](https://user-images.githubusercontent.com/15122230/188903642-b5afd9b7-a8d8-44a3-9f96-ac7a68ca777e.png)

About the condition, we compare rp id here just as what we do when dynamic allocation is enabled before. Here we just filter the executors which match the resource profile bound to a `taskSet`, and when assign tasks, `TaskSchedulerImpl` will further check the executor's available resources. cc @Ngone51

Reply:
The condition for dynamic allocation looks good to me. But the concern is for the static case:

```scala
(!dynamicEnabled && taskRp.isInstanceOf[TaskResourceProfile] &&
      executorRpId == ResourceProfile.DEFAULT_RESOURCE_PROFILE_ID)
```
The `executorRpId` here is still the rpId of `TaskResourceProfile`, right? But the rpId of `TaskResourceProfile` is not stick to the `ResourceProfile.DEFAULT_RESOURCE_PROFILE_ID` with the current implementation, which means `TaskResourceProfile` can't reuse the executors.